### PR TITLE
Restart jobs, tasks and allocs

### PR DESCRIPTION
**Pull request by cshintov**
Implements https://github.com/robinovitch61/wander/issues/120

Implemented:
	- Add restart task on key `t`
	- Load job tasks after restarting a task
TODO:
  - restart alloc
  - restart job

### CODE REVIEW

**Comment by robinovitch61** [comment]
Thanks for starting work on this @cshintov ! Check out some of my thoughts thoughts [here](https://github.com/robinovitch61/wander/issues/120#issuecomment-1902676389) on overall design for admin stuff and LMK what you think.

For this PR, consider incorporating some of that design - it would be cool if the "admin task" table had just a single entry for now, "Restart Job/Allocation <name>" by the end of this work. You can do jobs and allocations/tasks separately over different PRs if you'd like.

For each PR, target the `next` branch, as that's where new feature development should live.

**Comment by cshintov** [comment]
Continuing here.

> > Thanks for these thoughts. I like the X -> admin menu idea. Makes it explicit and cautions the users.
> > But I think the typing `yes` to do the action, a little tedious in a scenario like @attachmentgenie got into. I like the `k9s` way. Opening a pop up window, press tab to `cancel` or `ok` buttons, default can be `cancel`, and press enter to do the action.
> 
> If we can implement the pop up window nicely, that sounds great!


Looked into how this can be implemented. Not natively supported by lip gloss. It's in the pipeline.

There's a work around though!
https://github.com/charmbracelet/lipgloss/pull/102#issuecomment-1375875184

I think we can postpone this and go with your idea of typing `yes` for now. What do you think? Should I try to implement the pop up window like mentioned in the above comment instead?

**Comment by cshintov** [comment]
Btw, we can't restart a job. I Checked the api code. There's no such function. Probably why the UI also doesn't provide a button for this. 

When I want to restart a job, I usually add a dummy `meta` block to simulate/trigger a job restart. Maybe we can do [that](https://github.com/cshintov/nomad/blob/main/api/jobs.go#L1239) here as well.

**Comment by robinovitch61** [comment]
> Btw, we can't restart a job. I Checked the api code. There's no such function. Probably why the UI also doesn't provide a button for this. 
> 
> 
> 
> When I want to restart a job, I usually add a dummy `meta` block to simulate/trigger a job restart. Maybe we can do [that](https://github.com/cshintov/nomad/blob/main/api/jobs.go#L1239) here as well.

Interesting! Could we issue a stop, poll for when it's stopped, then issue a start? I'm afk so can't look at the nomad api right now.

Don't feel you have to implement more than one admin action in this first PR. I would pick the one that's easiest and add just that to the list of admin actions for now.

**Comment by cshintov** [comment]
Have implemented just restarting tasks for now, although `Admin Menu` lists others as well. Will remove.

The UI is not very nice, but does the job.

[wander-restart-task.webm](https://github.com/robinovitch61/wander/assets/11365900/61849d31-8158-4830-bb75-1497a0eb4c6c)

**Comment by robinovitch61** [comment]
> Have implemented just restarting tasks for now, although `Admin Menu` lists others as well. Will remove.
> 
> The UI is not very nice, but does the job.
> 
>  wander-restart-task.webm

Nice! I'll give the code a review this weekend :)

File: internal/dev/dev.go
Comment (robinovitch61) [inline internal/dev/dev.go:1]: I'd be open to modifying Debug to read an optional environment variable's value, `WANDER_LOG_PATH`, and falling back to `wander.log` if it's not set. That way we don't need multiple debug functions and devs can still specify their preferred logging path if not `wander.log`

File: README.md
Comment (robinovitch61) [inline README.md:1]: Woops, `next` wasn't up to date with `main`. Now it should be. Not sure why it's still showing this diff in the github ui

**Comment by robinovitch61** [comment]
Remember to modify `GetPageKeyHelp` to include the help text for `X admin`. I think we'll want it to the right of `ctrl+w toggle wrap` on the second row of the relevant pages.

File: internal/tui/nomad/pages.go [hunk 1]
```diff
@@ -314,6 +367,12 @@ func namespaceFilterPrefix(namespace string) string {
 
 func (p Page) GetFilterPrefix(namespace, jobID, taskName, allocName, allocID string, eventTopics Topics, eventNamespace string) string {
 	switch p {
+	case ConfirmPage:
+		return "🔴 Are you sure?"
```
Comment (robinovitch61) [inline internal/tui/nomad/pages.go:371]: I've found emojis to have weird width in terminals, e.g. https://github.com/charmbracelet/lipgloss/issues/55. Let's remove this one

**Comment by robinovitch61** [comment]
@cshintov I played around a bit here: https://github.com/robinovitch61/wander/pull/124

check it out and steal whatever you'd like to from there into your work here (I don't plan on merging my PR, it's just for sharing purposes)

File: internal/dev/dev.go
Comment (cshintov) [reply internal/dev/dev.go:1]: My intention was to rather redirect specific temporary debug logs to a second file for ease of parsing the logs. But the second function can be added ad hoc, no need to keep it in the code. Have removed it.

Btw, what do you use for your debugging, `delve` or `wander.log`?

**Comment by cshintov** [comment]
> @cshintov I played around a bit here: #124
> 
> check it out and steal whatever you'd like to from there into your work here (I don't plan on merging my PR, it's just for sharing purposes)

Thanks, that looks more clean, so I've stolen it all 😂! 

I've added the `X -> Admin` help and returning `FailureMsg` as you suggested.

**Comment by robinovitch61** [comment]
> > @cshintov I played around a bit here: #124
> > check it out and steal whatever you'd like to from there into your work here (I don't plan on merging my PR, it's just for sharing purposes)
> 
> Thanks, that looks more clean, so I've stolen it all 😂!
> 
> I've added the `X -> Admin` help and returning `FailureMsg` as you suggested.
> 
>  Screenshare.-.2024-01-28.11.10.00.PM.webm

Looks great! Will give this a final review today

File: internal/tui/components/app/app.go [hunk 2]
```diff
@@ -61,13 +63,15 @@ type Config struct {
 }
 
 type Model struct {
-	config Config
-	client api.Client
+	confirmed bool
```
Comment (robinovitch61) [inline internal/tui/components/app/app.go:66]:
```suggestion
```

think we don't need this now

File: internal/tui/components/app/app.go [hunk 3]
```diff
@@ -61,13 +63,15 @@ type Config struct {
...
+	client    api.Client
 
-	header      header.Model
-	compact     bool
-	currentPage nomad.Page
-	pageModels  map[nomad.Page]*page.Model
+	header       header.Model
+	compact      bool
+	currentPage  nomad.Page
+	previousPage nomad.Page
```
Comment (robinovitch61) [inline internal/tui/components/app/app.go:73]:
```suggestion
```

same here

File: internal/tui/components/app/app.go [hunk 4]
```diff
@@ -164,6 +171,7 @@ func (m Model) Update(msg tea.Msg) (tea.Model, tea.Cmd) {
 		}
 
 	case nomad.PageLoadedMsg:
+		dev.Debug(fmt.Sprintf("page loaded: %v", msg.Page))
```
Comment (robinovitch61) [inline internal/tui/components/app/app.go:174]:
```suggestion
```

let's remove the newly added dev.Debug lines

File: internal/tui/components/app/app.go [hunk 5]
```diff
@@ -423,9 +473,11 @@ func (m *Model) handleKeyMsg(msg tea.KeyMsg) tea.Cmd {
 
 		case key.Matches(msg, keymap.KeyMap.Reload):
 			if m.currentPage.DoesReload() {
+				m.getCurrentPageModel().HideToast()
```
Comment (robinovitch61) [inline internal/tui/components/app/app.go:476]: I actually don't think we want this - people might manually be refreshing the page every second or so manually with `r` and hide the toast too quickly

File: internal/tui/components/app/app.go [hunk 6]
```diff
@@ -644,18 +713,56 @@ func (m Model) getCurrentPageCmd() tea.Cmd {
...
+				TableHeader: []string{"Exec Session Output"},
+				AllPageRows: allPageRows,
+			}
 		}
 	case nomad.AllocSpecPage:
 		return nomad.FetchAllocSpec(m.client, m.alloc.ID)
 	case nomad.LogsPage:
-		return nomad.FetchLogs(m.client, m.alloc, m.taskName, m.logType, m.config.Log.Offset, m.config.Log.Tail)
+		return nomad.FetchLogs(
+			m.client, m.alloc, m.taskName, m.logType, m.config.Log.Offset, m.config.Log.Tail)
```
Comment (robinovitch61) [inline internal/tui/components/app/app.go:726]: Maybe your editor is automatically doing this reformatting, but if you don't mind undoing the formatting-only changes before merging, that would be great!

File: internal/tui/nomad/pages.go [hunk 7]
```diff
@@ -427,11 +473,23 @@ func GetPageKeyHelp(
 
 	viewportKeyMap := viewport.GetKeyMap()
 	secondRow := []key.Binding{viewportKeyMap.Save, keymap.KeyMap.Wrap}
-	thirdRow := []key.Binding{viewportKeyMap.Down, viewportKeyMap.Up, viewportKeyMap.PageDown, viewportKeyMap.PageUp, viewportKeyMap.Bottom, viewportKeyMap.Top}
+
+	if currentPage == JobTasksPage {
```
Comment (robinovitch61) [inline internal/tui/nomad/pages.go:477]: I think we'll want `AllTasksPage` to show this too - you can likely reuse `p.HasAdminMenu()` here

File: internal/tui/style/style.go [hunk 8]
```diff
@@ -43,4 +43,6 @@ var (
 	StdErr                        = Regular.Copy().Foreground(red)
 	SuccessToast                  = Bold.Copy().PaddingLeft(1).Foreground(black).Background(darkgreen)
 	ErrorToast                    = Bold.Copy().PaddingLeft(1).Foreground(black).Background(darkred)
+	Red                           = Regular.Copy().Foreground(red)
+	Green                         = Regular.Copy().Foreground(darkgreen)
```
Comment (robinovitch61) [inline internal/tui/style/style.go:47]:
```suggestion
```

Think we can remove too :)

**Review by robinovitch61 — APPROVED** [review]
Looks great! Left some final comments for you to consider, but feel free to merge whenever you feel good about it. Nice job, thanks so much for this!

**Comment by cshintov** [comment]
Made changes according to the review. Think it's ready now.

> but feel free to merge whenever you feel good about it.

But I don't have permission to merge!

**Pull request merged**